Backpack/OverlayView/Classes/BPKOverlayView.m

```diff
@@ -0,0 +1,192 @@
+
+    if (self) {
+        [self setupWithOverlayType:BPKOverlayViewOverlayTypeTint cornerStyle:BPKOverlayViewCornerStyleNone];
+    }
+
+    return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    BPKAssertMainThread();
+    self = [super initWithFrame:frame];
+
+    if (self) {
+        [self setupWithOverlayType:BPKOverlayViewOverlayTypeTint cornerStyle:BPKOverlayViewCornerStyleNone];
+    }
+
+    return self;
+}
+
+- (nullable instancetype)initWithOverlayType:(BPKOverlayViewOverlayType)overlayType cornerStyle:(BPKOverlayViewCornerStyle)cornerStyle {
+    BPKAssertMainThread();
+    self = [super initWithFrame:CGRectZero];
+
+    if (self) {
+        [self setupWithOverlayType:overlayType cornerStyle:cornerStyle];
+    }
+
+    return self;
+}
+
+- (void)layoutSubviews {
+    if (self.tintLayer == nil) {
+        return;
+    }
+
+    self.tintLayer.frame = self.bounds;
+}
+
+#pragma mark - Private
+
+- (void)setupWithOverlayType:(BPKOverlayViewOverlayType)overlayType cornerStyle:(BPKOverlayViewCornerStyle)cornerStyle {
+    self.overlayType = overlayType;
+    self.cornerStyle = cornerStyle;
+}
+
+#pragma mark - getters
+
+- (UIView *)foregroundView {
+    if (_foregroundView != nil) {
+        return _foregroundView;
+    }
+
+    _foregroundView = [[UIView alloc] initWithFrame:CGRectZero];
+    _foregroundView.translatesAutoresizingMaskIntoConstraints = NO;
+    [self addSubview:_foregroundView];
+
+    [NSLayoutConstraint activateConstraints:@[
+        [_foregroundView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+        [_foregroundView.topAnchor constraintEqualToAnchor:self.topAnchor],
+        [self.trailingAnchor constraintEqualToAnchor:_foregroundView.trailingAnchor],
+        [self.bottomAnchor constraintEqualToAnchor:_foregroundView.bottomAnchor]
+    ]];
+
+    return _foregroundView;
+}
+
+- (UIView *)backgroundView {
+    if (_backgroundView != nil) {
+        return _backgroundView;
+    }
+
+    self.tintLayer = [CALayer layer];
+    self.tintLayer.backgroundColor = BPKColor.skyBlueShade03.CGColor;
+    self.tintLayer.zPosition = 100;
+    self.layer.masksToBounds = YES;
+
+    _backgroundView = [[UIView alloc] initWithFrame:CGRectZero];
+    _backgroundView.translatesAutoresizingMaskIntoConstraints = NO;
+    [self insertSubview:_backgroundView atIndex:0];
+
+    [_backgroundView.layer addSublayer:self.tintLayer];
+    [self updateTintAppearance];
+
+    [NSLayoutConstraint activateConstraints:@[
+        [_backgroundView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+        [_backgroundView.topAnchor constraintEqualToAnchor:self.topAnchor],
+        [self.trailingAnchor constraintEqualToAnchor:_backgroundView.trailingAnchor],
+        [self.bottomAnchor constraintEqualToAnchor:_backgroundView.bottomAnchor]
+    ]];
+
+    return _backgroundView;
+}
+
+#pragma mark - setters
+
+- (void)setCornerStyle:(BPKOverlayViewCornerStyle)cornerStyle {
```

Comment:
This and several other methods are missing `BPKAssertMainThread`

Reply:
Fixed. AFAIK we should use these on all initialisers and setters, right?

Reply:
Yeah all the ones that end up affect UI properties such as layer styles